gRPC large binary performance best practices (#24250)
dotnet/AspNetCore.Docs: aspnetcore/grpc/performance.md

* Attempts to get an array from `ByteString.Memory` with <xref:System.Runtime.InteropServices.MemoryMarshal.TryGetArray%2A?displayProperty=nameWithType>.
* Uses the `ArraySegment<byte>` if it was successfully retrieved. The segment has a reference to the array, offset and count.
* Otherwise, falls back to allocating a new array with `ByteString.ToByteArray()`.

### gRPC services and large binary payloads

> gRPC and Protobuf can send and receive large binary payloads. Although binary Protobuf is more efficent than text-based JSON at serializing binary payloads, there are still important performance characteristics to keep in mind when working with large binary payloads.

Reviewer: ```suggestion
gRPC and Protobuf can send and receive large binary payloads. Although binary Protobuf is more efficient than text-based JSON at serializing binary payloads, there are still important performance characteristics to keep in mind when working with large binary payloads.
```

> gRPC is a message based RPC framework. That means that the entire message is loaded into memory before gRPC can send it, and the entire message is deserialized into memory when received. Messages with large binary payloads can allocate byte arrays on the [large object heap](/dotnet/standard/garbage-collection/large-object-heap). Large allocations impact server performance and scalablity.

Reviewer: ```suggestion
gRPC is a message based RPC framework. That means that the entire message is loaded into memory before gRPC can send it, and the entire message is deserialized into memory when received. Messages with large binary payloads can allocate byte arrays on the [large object heap](/dotnet/standard/garbage-collection/large-object-heap). Large allocations impact server performance and scalability.
```

> gRPC is a message based RPC framework. That means that the entire message is loaded into memory before gRPC can send it, and the entire message is deserialized into memory when received. Messages with large binary payloads can allocate byte arrays on the [large object heap](/dotnet/standard/garbage-collection/large-object-heap). Large allocations impact server performance and scalability.

Reviewer 2: ```suggestion
gRPC is a message-based RPC framework, which means:

* The entire message is loaded into memory before gRPC can send it.
* When the message is received, the entire message is deserialized into memory.

Messages with large binary payloads can allocate byte arrays on the [large object heap](/dotnet/standard/garbage-collection/large-object-heap). Large allocations impact server performance and scalability.
```

Advice for creating high-performance applications with large binary payloads:

> * **Avoid** very large binary payloads in gRPC messages.

Reviewer 2: ```suggestion
* **Avoid** large binary payloads in gRPC messages.
```

I'm not sure if it says anywhere what size a typical "large" binary payload is. This could state a typical value ...

> \*\*Avoid\*\* large binary payloads in gRPC messages, such as payloads that exceed approximately XXXXX MB.

Reviewer: How about greater than 20% available memory? Hard to put a number on it when VMs memory size varies by such a large factor

> * **Consider** splitting up large binary payloads [using streaming](xref:grpc/client#client-streaming-call). Binary data is chunked and streamed over multiple messages.

Reviewer 2: ```suggestion
* **Consider** splitting large binary payloads [using streaming](xref:grpc/client#client-streaming-call). Binary data is chunked and streamed over multiple messages.
```

> * **Consider** not using gRPC for large binary data. HTTP endpoints can be used alongside gRPC services. An HTTP endpoint that supports sending or receiving large files using the stream body is very efficient.

Reviewer: ```suggestion
* **Consider** not using gRPC for large binary data. HTTP endpoints can be used alongside gRPC services. An HTTP endpoint that supports sending or receiving large files using the stream body is efficient.
```

* Attempts to get an array from `ByteString.Memory` with <xref:System.Runtime.InteropServices.MemoryMarshal.TryGetArray%2A?displayProperty=nameWithType>.
* Uses the `ArraySegment<byte>` if it was successfully retrieved. The segment has a reference to the array, offset and count.
* Otherwise, falls back to allocating a new array with `ByteString.ToByteArray()`.

### gRPC services and large binary payloads

gRPC and Protobuf can send and receive large binary payloads. Although binary Protobuf is more efficient than text-based JSON at serializing binary payloads, there are still important performance characteristics to keep in mind when working with large binary payloads.

gRPC is a message-based RPC framework, which means:

* The entire message is loaded into memory before gRPC can send it.
* When the message is received, the entire message is deserialized into memory.

Messages with large binary payloads can allocate byte arrays on the [large object heap](/dotnet/standard/garbage-collection/large-object-heap). Large allocations impact server performance and scalability.

Advice for creating high-performance applications with large binary payloads:

> * **Avoid** very large binary payloads in gRPC messages.

Author: ```suggestion
* **Avoid** large binary payloads in gRPC messages. Arrays larger than 85,000 bytes are allocated on the large object heap.
```
